### PR TITLE
Added a 'from_hsl' function to core rgba.cr

### PR DESCRIPTION
**Pull request by Demonstrandum**
The `from_hsl` function allows writing to an image through the HSL colour format, it converts the HSL value in to RGB values and uses a bit-depth of 8 using the preexisting `from_rgb_n(r, g, b, 8)` function.
Here's an example using the modified stumpy_core rgba.cr file:
```ruby
require "stumpy_png"
include StumpyPNG

spectrum = Canvas.new(361, 101)
(0..360.each do |x|
  (0..100).each do |y|
    # RGBA.from_hsla_n(hue, saturation, lightness, alpha, bit_depth) is an internal helper method
    color = RGBA.from_hsla_n(x, 100, y, UInt16::MAX, 8) # Here from_hsl(x, 100, y) would work exactly the same
    spectrum[x, y] = color
  end
end

StumpyPNG.write(spectrum, "spectrum.png")
```
Which produces:
spectrum

`h` (hue) is in degrees (from 0°..360°) and
`s` (saturation) & `l` (lightness) are in percent, (0%..100%)

### CODE REVIEW

File: src/stumpy_core/rgba.cr [hunk 1]
```diff
@@ -120,6 +120,32 @@ module StumpyCore
...
+          return a
+        end
+        b = (l < 0.5) ? (l * (1.0 + s)) : (l + s - l * s)
+        a = 2.0 * l - b
+        r = hue.call(a, b, h + 1.0/3.0)
+        g = hue.call(a, b, h)
+        b = hue.call(a, b, h - 1.0/3.0)
+      end
+
+      return from_rgb_n((r * 255.0).round.to_i, (g * 255.0).round.to_i, (b * 255.0).round.to_i, 8)
```
Comment (l3kn) [inline src/stumpy_core/rgba.cr]: You could use `from_relative(r, g, b, 1.0)` here

File: src/stumpy_core/rgba.cr [hunk 2]
```diff
@@ -120,6 +120,32 @@ module StumpyCore
...
 
+    def self.from_hsl(h, s, l)
+      h /= 360.0
+      s /= 100.0
+      l /= 100.0
+
+      if s == 0
+        r = g = b = l
+      else
+        hue = Proc(Float64, Float64, Float64, Float64).new do | a, b, c |
```
Comment (l3kn) [inline src/stumpy_core/rgba.cr]: Why not make this a private helper function like `get_hsl_hue`?

File: src/stumpy_core/rgba.cr [hunk 3]
```diff
@@ -120,6 +120,32 @@ module StumpyCore
       to_rgb8
     end
 
+    def self.from_hsl(h, s, l)
+      h /= 360.0
+      s /= 100.0
+      l /= 100.0
+
```
Comment (l3kn) [inline src/stumpy_core/rgba.cr]: For `from_relative` to work, `r, g, b` must be floats
and I like initializing them unconditionally better

``` crystal
r = g = b = 1.0

if s != 0
...
```

**Review by l3kn — COMMENTED** [review]
Thanks for the pull request!

I added some comments
to the parts of the code I'm not so sure about

There is some minor stuff to do
like increasing the version in `shard.yml` to `1.4.1`
and documenting the function,
but I can do that later, there is some other stuff missing
in the docs ;)

**Comment by Demonstrandum** [comment]
Updated code for suggestions, added more HSL type functions (from_hsla_n, from_hsla, from_hsl_n, from_hsl, etc…) (also works with tuples/arrays), increased version and added information to README

**Comment by l3kn** [comment]
Nice!

What would be the use case for `from_hsla_n` etc?
The `from_rgba_n`, ... functions could be used when working with image formats
that enocde colors in 1bit / 2bit / 4bit / 8bit per channel,
but from what I've seen hsl uses either (0..360, 0..100, 0..100)
or (0..1, 0..1, 0..1) for the values.

**Comment by Demonstrandum** [comment]
That's true! bit-depth would be completely unnecessary using HSL as it will always be between `0..360` and `0..100`. RGB values would always have the potential to be between 0..255 using this method. I'll remove the functions with options for bit-depth.

**Pull request merged**